chapter-06/inprogress/suggestions.go

```diff
@@ -0,0 +1,106 @@
+package main
+
+import "sort"
+
+// bookSuggestions are the suggested books for a given book.
+// Suggestions are the neighbour-books on the shelves.
+type bookSuggestions map[Book]bookCollection
+
+// bookCollection is a collection of books.
+type bookCollection map[Book]struct{}
+
+// suggestOtherBooks returns the suggestions book from the matching-book.
+func suggestOtherBooks(bookworms []Bookworm) []Bookworm {
+	sb := make(bookSuggestions)
+
+	// Register all books with their suggestions meaning the others books on the same shelves' owner.
+	for _, bookworm := range bookworms {
+		for i, book := range bookworm.Books {
+			otherBooksOnShelves := buildOtherBooksOnShelves(i, bookworm.Books)
+			registerBookSuggestions(sb, book, otherBooksOnShelves)
+		}
+	}
+
+	// Suggest a list of related books to each bookworm.
+	suggestions := make([]Bookworm, len(bookworms))
+	for i, bookworm := range bookworms {
+		suggestions[i] = Bookworm{
+			Name:  bookworm.Name,
+			Books: findSuggestionsBook(sb, bookworm.Books),
+		}
+	}
+
+	return suggestions
```

Comment:
This slice is sorted in the order of the input bookworms slice. It's worth mentioning it somewhere - probably in the book.